### PR TITLE
build(deps): vendor libvterm at v0.3.3 and enable reflow (for realsies this time)

### PR DESCRIPTION
**Pull request by clason**
Problem: Adding support for modern Nvim features (reflow, OSC 8, full utf8/emoji support) requires coupling libvterm to Nvim internals (e.g., utf8proc).

Solution: Vendor libvterm at v0.3.3. As a first step, make necessary fixes to enable terminal reflow again.

Question: How much should we invest in improving code quality? (Besides a lot of stylistic warnings, clangd raises a few valid complaints.)

Follow-ups:
* check recent commits on https://github.com/neovim/libvterm/commits/mirror/ for possible cherry-picking
* refactor duplicated code from `libtermkey` (but note different pointer arguments)
* replace `unicode.c` and `fullwidth.inc` (and `utf8.h`) with `mbyte.c`/`utf8proc` functions
* convert upstream tests to `unittest`s

### CODE REVIEW

File: src/vterm/vterm.h [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+    unsigned int underline : 2;
+    unsigned int italic    : 1;
+    unsigned int blink     : 1;
+    unsigned int reverse   : 1;
+    unsigned int conceal   : 1;
+    unsigned int strike    : 1;
+    unsigned int font      : 4; /* 0 to 9 */
+    unsigned int dwl       : 1; /* On a DECDWL or DECDHL line */
+    unsigned int dhl       : 2; /* On a DECDHL line (1=top 2=bottom) */
+    unsigned int small     : 1;
```
Comment (clason) [inline src/vterm/vterm.h:510]: looks like MSVC doesn't like this sort of struct initialization?

Comment (clason) [reply src/vterm/vterm.h:510]: Or the inline comments?

Comment (clason) [reply src/vterm/vterm.h:510]: Or the artificial alignment?

File: src/vterm/utf8.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/* The following functions copied and adapted from libtermkey
```
Comment (clason) [inline src/vterm/utf8.h:1]: Code duplication: if we vendor both libs, we might want to consolidate.

File: src/nvim/terminal.c [hunk 3]
```diff
@@ -320,7 +320,7 @@ void terminal_open(Terminal **termpp, buf_T *buf, TerminalOptions opts)
   term->vts = vterm_obtain_screen(term->vt);
   vterm_screen_enable_altscreen(term->vts, true);
   // TODO(clason): reenable when https://github.com/neovim/neovim/issues/23762 is fixed
```
Comment (clason) [inline src/nvim/terminal.c:322]: @zeertzjq I've pulled in the fixes suggested in https://github.com/neovim/neovim/issues/23762 and haven't seen any issues so far. I'd appreciate it if you could test your reproducers and see if they still fail.

File: src/vterm/encoding/DECdrawing.tbl [hunk 4]
```diff
@@ -0,0 +1,31 @@
+6/0 = U+25C6 # BLACK DIAMOND
```
Comment (clason) [inline src/vterm/encoding/DECdrawing.tbl:1]: Question: should we vendor the tables as well? Or just move the comments to the include files?

(Longterm we probably want to replace all of this with `utf8proc`?)

File: src/vterm/unicode.c [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+  if(bisearch(codepoint, fullwidth, sizeof(fullwidth) / sizeof(fullwidth[0]) - 1))
+    return 2;
+
+  return mk_wcwidth(codepoint);
+}
+
+INTERNAL int vterm_unicode_is_combining(uint32_t codepoint)
+{
+  return bisearch(codepoint, combining, sizeof(combining) / sizeof(struct interval) - 1);
+}
```
Comment (clason) [inline src/vterm/unicode.c:313]: Same for these: just use `utf8proc` instead in `state.c`?

File: src/vterm/vterm.h (hunk 1 above)
Comment (dundargoc) [reply src/vterm/vterm.h:510]: `<windows.h>` includes `<rpcndr.h>`, which contains the following:

```
#define small char
```

Adding a `#undef small` after `#include <windows.h>` in `src/nvim/os/win_defs.h` removes the warning.

Comment (clason) [reply src/vterm/vterm.h:510]: Do we want this, or just rename `small` to `little` or `tiny`? :P

**Comment by dundargoc** [comment]
There is a `# libvterm-dev` comment in `.github/workflows/test.yml` you can remove.

File: src/vterm/unicode.c (hunk 5 above)
Comment (clason) [reply src/vterm/unicode.c:313]: (`wcwidth` at home...)

File: runtime/doc/news.txt [hunk 6]
```diff
@@ -163,6 +163,9 @@ TERMINAL
   system clipboard (copy). Querying with OSC 52 (paste) is not supported.
 • |hl-StatusLineTerm| and |hl-StatusLineTermNC| define highlights for the
   status line in |terminal| windows.
+• The terminal buffer now supports reflow (wrapped lines adapt when the buffer
+  is resized horizontally). Note: Lines that are not visible and kept in
+  |'scrollback'| are not reflown.
```
Comment (zeertzjq) [inline runtime/doc/news.txt:168]:
```suggestion
  'scrollback' are not reflown.
```

File: src/nvim/terminal.c (hunk 3 above)
Comment (zeertzjq) [reply src/nvim/terminal.c:322]: I can't reproduce the heap-buffer-overflow on this branch.
```suggestion
```

File: runtime/doc/news.txt (hunk 6 above)
Comment (clason) [reply runtime/doc/news.txt:168]: Why? I was explicitly asked to link this, IIRC.

Comment (zeertzjq) [reply runtime/doc/news.txt:168]: You shouldn't use `|` to link to an option. The single quotes are enough.

**Comment by clason** [comment]
It's all yours, @gpanders !

**Pull request merged**